stan/math/rev/functor/ad_stack_matrix.hpp

```diff
@@ -0,0 +1,111 @@
+#ifndef STAN_MATH_REV_FUNCTOR_AD_STACK_MATRIX_HPP
+#define STAN_MATH_REV_FUNCTOR_AD_STACK_MATRIX_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/rev/core.hpp>
+
+namespace stan {
+namespace math {
+
+template <typename MatrixType>
+class AD_stack_matrix : public Eigen::Map<MatrixType> {
+ public:
+  using Scalar = value_type_t<MatrixType>;
+  static constexpr int RowsAtCompileTime = MatrixType::RowsAtCompileTime;
+  static constexpr int ColsAtCompileTime = MatrixType::ColsAtCompileTime;
+
+  /**
+   * Default constructor.
+   */
+  AD_stack_matrix()
+      : Eigen::Map<MatrixType>::Map(
+            nullptr,
+            RowsAtCompileTime == Eigen::Dynamic ? 0 : RowsAtCompileTime,
+            ColsAtCompileTime == Eigen::Dynamic ? 0 : ColsAtCompileTime) {}
+
+  /**
+   * Constructs `AD_stack_matrix` with given number of rows and columns.
+   * @param rows number of rows
+   * @param cols number of columns
+   */
+  AD_stack_matrix(Eigen::Index rows, Eigen::Index cols)
+      : Eigen::Map<MatrixType>::Map(
+            ChainableStack::instance_->memalloc_.alloc_array<Scalar>(rows
+                                                                     * cols),
+            rows, cols) {}
+
+  /**
+   * Constructs `AD_stack_matrix` with given size. This only works if
+   * `MatrixType` is row or col vector.
+   * @param size number of elements
+   */
+  explicit AD_stack_matrix(Eigen::Index size)
+      : Eigen::Map<MatrixType>::Map(
+            ChainableStack::instance_->memalloc_.alloc_array<Scalar>(size),
+            size) {}
+
+  /**
+   * Constructs `AD_stack_matrix` form an expression.
+   * @param other expression
+   */
+  template <typename T, require_eigen_t<T>* = nullptr>
+  AD_stack_matrix(const T& other)  // NOLINT
+      : Eigen::Map<MatrixType>::Map(
+            ChainableStack::instance_->memalloc_.alloc_array<Scalar>(
+                other.size()),
+            other.rows(), other.cols()) {
+    *this = other;
+  }
+
+  /**
+   * Copy constructor.
+   * @param other matrix to copy from
+   */
+  AD_stack_matrix(const AD_stack_matrix<MatrixType>& other)
+      : Eigen::Map<MatrixType>::Map(const_cast<Scalar*>(other.data()),
+                                    other.rows(), other.cols()) {}
+
+  using Eigen::Map<MatrixType>::operator=;
+
+  /**
+   * Copy assignment operator.
+   * @param other matrix to copy from
+   * @return `*this`
+   */
+  AD_stack_matrix& operator=(const AD_stack_matrix<MatrixType>& other) {
+    // placement new changes what data map points to - there is no allocation
+    new (this) Eigen::Map<MatrixType>(const_cast<Scalar*>(other.data()),
+                                      other.rows(), other.cols());
+    return *this;
+  }
+
+  /**
+   * Resizes this object to given number of rows and columns.
+   * @param rows number of rows
+   * @param cols number of columns
+   */
+  void resize(Eigen::Index rows, Eigen::Index cols) {
```

Comment:
This won't work for const things

Reply:
It does not need to. Const objects are const and cant be resized.

Reply:
Oh resize reallocates too. I don't see where I would resize anything though. Am I missing an obvious use here?

Like `stack_matrix` just kinda wraps stuff that's already been allocated and math has been done on it and all.

Reply:
Resize is used by Eigen when a default constructed object is assigned an expression. However, it does not call the derived version, so I had to override the assignment operator to resize before delegating to base operator. So this is not actually needed. Deleted.